src/fs.ts

```diff
@@ -0,0 +1,258 @@
+// This file is intended to be used only in a Node.js context.
+/* eslint-disable import/no-nodejs-modules */
+
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+
+import { isErrorWithCode, wrapError } from './errors';
+import type { Json } from './json';
+
+/**
+ * Information about the file sandbox provided to tests that need temporary
+ * access to the filesystem.
+ */
+export type FileSandbox = {
+  directoryPath: string;
+  withinSandbox: (
+    test: (args: { directoryPath: string }) => Promise<void>,
+  ) => Promise<void>;
+};
+
+/**
+ * Reads the file at the given path, assuming its content is encoded as UTF-8.
+ *
+ * @param filePath - The path to the file.
+ * @returns The content of the file.
+ * @throws An error with a stack trace if reading fails in any way.
+ */
+export async function readFile(filePath: string): Promise<string> {
+  try {
+    return await fs.promises.readFile(filePath, 'utf8');
+  } catch (error) {
+    throw wrapError(error, `Could not read file '${filePath}'`);
+  }
+}
+
+/**
+ * Writes content to the file at the given path.
+ *
+ * @param filePath - The path to the file.
+ * @param content - The new content of the file.
+ * @throws An error with a stack trace if writing fails in any way.
+ */
+export async function writeFile(
+  filePath: string,
+  content: string,
+): Promise<void> {
+  try {
+    await fs.promises.mkdir(path.dirname(filePath), { recursive: true });
+    await fs.promises.writeFile(filePath, content);
+  } catch (error) {
+    throw wrapError(error, `Could not write file '${filePath}'`);
+  }
+}
+
+/**
+ * Reads the assumed JSON file at the given path, attempts to parse it, and
+ * returns the resulting object. Supports a custom parser (in case you want to
+ * use the [JSON5](https://www.npmjs.com/package/json5) package instead).
+ *
+ * @param filePath - The path segments pointing to the JSON file. Will be passed
+ * to path.join().
+ * @param options - Options to this function.
+ * @param options.parser - The parser object to use. Defaults to `JSON`.
+ * @param options.parser.parse - A function that parses JSON data.
+ * @returns The object corresponding to the parsed JSON file, typed against the
+ * struct.
+ * @throws An error with a stack trace if reading fails in any way, or if the
+ * parsed value is not a plain object.
+ */
+export async function readJsonFile<Value extends Json>(
+  filePath: string,
+  {
+    parser = JSON,
+  }: {
+    parser?: { parse: typeof JSON.parse };
```

Comment:
The signature for `JSON.parse` is:

```
parse(text: string, reviver?: (this: any, key: string, value: any) => any): any;
```

but the `reviver` is not used. It might be better to use something like `(text: string) => any`.

Reply:
Sure. I did want to make sure we reference `JSON.parse` but I've simplified that here: [8872b26](https://github.com/MetaMask/utils/pull/148/commits/8872b267432254d33f6db1fa794c0a90a3d84127)